server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java

```diff
@@ -75,23 +75,25 @@ private static String toLoggerName(final Class<?> cls) {
      * Logs a message at the {@link #DEPRECATION} level. The message is also sent to the header warning logger,
      * so that it can be returned to the client.
      */
-    public DeprecationLoggerBuilder deprecate(
+    public DeprecationLogger deprecate(
         final DeprecationCategory category,
         final String key,
         final String msg,
         final Object... params
     ) {
-        return new DeprecationLoggerBuilder().withDeprecation(category, key, msg, params);
+        ESLogMessage deprecationMessage = DeprecatedMessage.of(category, key, HeaderWarning.getXOpaqueId(), msg, params);
+        logger.log(DEPRECATION, deprecationMessage);
+        return this;
     }
 
-    public class DeprecationLoggerBuilder {
-
-        public DeprecationLoggerBuilder withDeprecation(DeprecationCategory category, String key, String msg, Object[] params) {
-            ESLogMessage deprecationMessage = DeprecatedMessage.of(category, key, HeaderWarning.getXOpaqueId(), msg, params);
-
-            logger.log(DEPRECATION, deprecationMessage);
-
-            return this;
-        }
+    public DeprecationLogger compatibleApiWarning(
```

Comment:
I wonder if we should leave this class as-is and require the user to pass the CompatibleApi category. Now that we are re-using this log and logger, it feels abit off for this to be special enough to warrant it's own method.

Reply:
I think the idea with the builder was over complicated. It was meant to allow for constructing some more complex DSLs (like emitting a compatible warning only when a deprecation warning was already called etc)
I think we don't need this, and also I am worried that we would unnecessarily create objects.

Regarding the separate method. I think it would be beneficial to have it. We hide from a user (a developer) that he has to remember about the `Category` field.

Reply:
+1 on the simplification, thanks for the explanation. 

re: separate method, I worry abit about the perception this is special in a way that it is not. It introduces a different way from everything. A dev adding a compatible warning will need to know they are adding a compatible warning, and this is mostly about which keys they need type to do that.  I would prefer to prevent differences without substance.

Reply:
I am not sure I understand. The intention of a separate method is to actually hide that additional category key from a developer. Using just a category field would mean that a developer would have to remember to use it, it would be an unnecessary duplication. 
I might disagree with just using a deprecation log method, but I don't think it is all that relevant. There is no usage of it now and this PR was just meant to introduce an infrastructure. 
In fact, if the method is to be removed, there is actually no point for this PR at all.

Reply:
OK ... my argument was that we shouldn't try to hide the additional category.  However, I am fine with the additional method.  Might want to consider adding an assert so that this method is the only way to set that category.